src/ngx_http_lua_kong_module.c

```diff
@@ -305,3 +305,172 @@ ngx_http_lua_kong_ffi_get_full_client_certificate_chain(ngx_http_request_t *r,
+                   "overriding upstream SSL configuration");
+
+    chain = ctx->upstream_client_certificate_chain;
+    pkey = ctx->upstream_client_private_key;
+
+    if (sk_X509_num(chain) < 1) {
+        ngx_ssl_error(NGX_LOG_ALERT, c->log, 0,
+                      "invalid client certificate chain provided while "
+                      "handshaking with upstream");
+        goto failed;
+    }
+
+    x509 = sk_X509_value(chain, 0);
+    if (x509 == NULL) {
+        ngx_ssl_error(NGX_LOG_ALERT, c->log, 0, "sk_X509_value() failed");
+        goto failed;
+    }
+
+    if (SSL_use_certificate(sc, x509) == 0) {
+        ngx_ssl_error(NGX_LOG_ALERT, c->log, 0, "SSL_use_certificate() failed");
+        goto failed;
+    }
+
+    /* read rest of the chain */
+
+    for (i = 1; i < sk_X509_num(chain); i++) {
+        x509 = sk_X509_value(chain, i);
+        if (x509 == NULL) {
+            ngx_ssl_error(NGX_LOG_ALERT, c->log, 0, "sk_X509_value() failed");
+            goto failed;
+        }
+
+        if (SSL_add1_chain_cert(sc, x509) == 0) {
+            ngx_ssl_error(NGX_LOG_ALERT, c->log, 0, "SSL_add1_chain_cert() failed");
+            goto failed;
+        }
+    }
+
+    if (SSL_use_PrivateKey(sc, pkey) == 0) {
+        ngx_ssl_error(NGX_LOG_ALERT, c->log, 0, "SSL_use_PrivateKey() failed");
+        goto failed;
+    }
+
+    return;
+
+failed:
+
+    ERR_clear_error();
+}
+
+
+static X509 *
+ngx_http_lua_kong_x509_copy(X509 *in)
+{
+    return X509_up_ref(in) == 0 ? NULL : in;
+}
+
+
+static void
+ngx_http_lua_kong_cleanup(void *data)
+{
+    ngx_http_lua_kong_ctx_t     *ctx = data;
+
+    if (ctx->upstream_client_certificate_chain) {
+        sk_X509_pop_free(ctx->upstream_client_certificate_chain, X509_free);
+        EVP_PKEY_free(ctx->upstream_client_private_key);
+    }
+}
+
+
+int
+ngx_http_lua_kong_ffi_set_upstream_client_cert_and_key(ngx_http_request_t *r,
+    void *_chain, void *_key)
+{
+    STACK_OF(X509)              *chain = _chain;
+    EVP_PKEY                    *key = _key;
+    STACK_OF(X509)              *new_chain;
+    ngx_http_lua_kong_ctx_t     *ctx;
+    ngx_pool_cleanup_t          *cln;
+
+    ctx = ngx_http_get_module_ctx(r, ngx_http_lua_kong_module);
+    if (ctx == NULL) {
+        ctx = ngx_pcalloc(r->pool, sizeof(ngx_http_lua_kong_ctx_t));
+        if (ctx == NULL) {
+            return NGX_ERROR;
+        }
+
+        cln = ngx_pool_cleanup_add(r->pool, 0);
+        if (cln == NULL) {
+            return NGX_ERROR;
+        }
+
+        cln->data = ctx;
+        cln->handler = ngx_http_lua_kong_cleanup;
+
+        ngx_http_set_ctx(r, ctx, ngx_http_lua_kong_module);
+
+    } else {
+        if (ctx->upstream_client_certificate_chain != NULL) {
+            ngx_http_lua_kong_cleanup(ctx);
+
+            ctx->upstream_client_certificate_chain = NULL;
+            ctx->upstream_client_private_key = NULL;
+        }
+    }
+
+    if (chain == NULL || key == NULL) {
+        return NGX_ERROR;
+    }
```

Comment:
It sounds to me like we could move this branch all the way up in this function's body? Or is the intent to clear our `ngx_http_lua_kong_ctx_t` even when the arguments to this function are invalid (which may sound like an undesired side-effect)? In any case, this shouldn't happen due to our safeguards in the Lua-land, but it does make the reader's life a bit easier if it could be moved up, I would say.
I am fine either way, just bringing it to your attention just in case.

Reply:
Good catch. Changed.